lib/Containers/MerkleTree.cpp

```diff
@@ -200,14 +227,106 @@ MerkleTree<Hasher, BranchingBits, LockStripes>::fromBuffer(std::string_view buff
+  size_t length;
+  bool canUncompress = snappy::GetUncompressedLength(buffer.data(), buffer.size(), &length);
+  if (!canUncompress || length != allocationSize(6)) {
+    throw std::invalid_argument(
+        "Cannot determine size of Snappy-compressed data.");
+  }
+  
+  auto uncompressed = std::make_unique<uint8_t[]>(length);
+
+  if (!snappy::RawUncompress(buffer.data(), buffer.size(), reinterpret_cast<char*>(uncompressed.get()))) {
+    throw std::invalid_argument(
+        "Cannot uncompress Snappy-compressed data.");
+  }
+  
+  return std::unique_ptr<MerkleTree<Hasher, BranchingBits>>(
+      new MerkleTree<Hasher, BranchingBits>(std::move(uncompressed)));
+}
+
+template <typename Hasher, std::uint64_t const BranchingBits>
+std::unique_ptr<MerkleTree<Hasher, BranchingBits>>
+MerkleTree<Hasher, BranchingBits>::fromBottomMostCompressed(std::string_view buffer) {
+  char const* p = buffer.data();
+  char const* e = p + buffer.size();
+
+  if (p + 3 * sizeof(uint64_t) > e) {
+    throw std::invalid_argument("invalid compressed tree data");
+  }
+
+  std::uint64_t rangeMin = readUInt<uint64_t>(p);
+  std::uint64_t rangeMax = readUInt<uint64_t>(p);
+  std::uint64_t maxDepth = readUInt<uint64_t>(p);
+
+  auto tree = std::make_unique<MerkleTree<Hasher, BranchingBits>>(maxDepth,
+                                                                  rangeMin,
+                                                                  rangeMax);
+  if (tree == nullptr) {
+    throw std::invalid_argument("invalid compressed tree data");
+  }
+
+  std::uint64_t const offset = nodeCountUpToDepth(maxDepth - 1);
+
+  // rebuild bottom-most level
+  while (p + sizeof(uint32_t) + sizeof(uint64_t) + sizeof(uint64_t) <= e) {
+    // enough data to read
+    std::uint32_t pos = readUInt<uint32_t>(p);
+    std::uint64_t count = readUInt<uint64_t>(p);
+    std::uint64_t hash = readUInt<uint64_t>(p);
+
+    Node& node = tree->node(offset + pos);
+    node.count = count;
+    node.hash = hash;
+  }
+
+  if (p != e) {
+    throw std::invalid_argument("invalid compressed tree data with overflow values");
+  }
+  
+  // rebuild all levels above from bottom-most data
+  for (std::size_t d = maxDepth - 1; /*d >= 0*/; --d) {
+    std::uint64_t prevOffset = (d == 0 ? 0 : nodeCountUpToDepth(d - 1));
+    std::uint64_t ourOffset = nodeCountUpToDepth(d);
+    std::uint64_t const ourEnd = nodeCountUpToDepth(d + 1);
+
+    while (ourOffset < ourEnd) {
+      std::uint64_t count = 0;
+      std::uint64_t hash = 0;
+      for (std::size_t index = 0; index < (1 << BranchingBits); ++index) {
+        Node const& src = tree->node(ourOffset++);
+        count += src.count;
+        hash ^= src.hash;
+      }
+      Node& dst = tree->node(prevOffset);
+      dst.count = count;
+      dst.hash = hash;
+
+      ++prevOffset;
+    }
+    if (d == 0) { 
+      break;
+    }
+  }
+
+#ifdef PARANOID_TREE_CHECKS
+  tree->checkInternalConsistency();
```

Comment:
Can we enable this for maintainer mode in general?

Reply:
Yes, we could, but this will make tests a lot slower, and I am not sure if we want to do it or not.